python_modules/dagster/dagster/_core/ext/context.py

```diff
@@ -43,10 +55,9 @@ def _handle_report_asset_metadata(
         check.str_param(asset_key, "asset_key")
         check.str_param(label, "label")
         check.opt_literal_param(type, "type", get_args(ExtMetadataType))
-        key = AssetKey.from_user_string(asset_key)
-        output_name = self._context.output_for_asset_key(key)
+        resolved_asset_key = AssetKey.from_user_string(asset_key)
         metadata_value = self._resolve_metadata_value(value, type)
-        self._context.add_output_metadata({label: metadata_value}, output_name)
+        self._metadata.setdefault(resolved_asset_key, {})[label] = metadata_value
```

Comment:
prefer a not mutative version of this

Reply:
As in reassigning the whole of `self._metadata` whenever a new value is added?

```
self._metadata = {
    **self._metadata,
    resolved_asset_key: {
        **self._metadata.get(resolved_asset_key, {}),
        label: metadata_value
    }
}
```

IMO this is a lot harder to parse and mutation is appropriate here since we are buffering values as they come in.

Reply:
I would prefer more explicit and less clever code

```python
if resolved_asset_key not in self._metadata:
    self._metadata[resolved_asset_key] = {}

self._metadata[resolved_asset_key][label] = metadata_value
```

rather than what is currently there:

```python
self._metadata.setdefault(resolved_asset_key, {})[label] = metadata_value
```

which is less code but hurts my brain